util/quotapool: enhance QuotaPool for use as a rate limiter

ajwerner opened the pull request:
This set of commits precedes work on #47906 where we want to provide rate-limiting. As a first step it extends the `QuotaPool` abstraction to be usable additionally for token-bucket style rate-limiting. This rate limiter differs from many token buckets as previously acquired quota can be returned in the case that the request cannot proceed.

This PR adds an example implementation of a rate limiter and basic testing.

There's a certain sense in which this is adding potentially more cruft to a potentially already cruft API. I'd love to hear more feedback there. Most of the cruft as I see it has to do with the attempts to avoid allocations. On the whole I'm reasonably happy with how these seems to feel as I move forward with the implementation of the tenant rate-limiter and its associated metrics. The hooks that already exist in the quotapool are quite handy.

Commit:
util/quotapool: adapt quotapool for use as a token-bucket rate-limiter

This commit extends the `QuotaPool` construct to allow for implementations
which act as rate-limiters. A rate-limiter allows a certain amount of
"quota" to be consumed per unit time. In general, when a rate-limited
request completes, it does not "release" any acquired quota back into
the pool.

Implementing a rate-limiter like this is handy because it provides us with
the possibility of releasing previously acquire quota back into the pool
if we discover that a request, for whatever reason, cannot proceed.

This turns out to be handy later on.

Release note: None

Commit:
util/quotapool: introduce TimeSource which will be utilized for testing

Release note: None

Commit:
util/quotapool: add ManualTime TimeSource implementation

Release note: None

Commit:
util/quotapool: add test of LogSlowAcquisition utilizing the TimeSource

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50340)
<!-- Reviewable:end -->

nvb reviewed (commented):
> Reviewed 3 of 3 files at r1, 2 of 2 files at r2, 1 of 1 files at r3, 2 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50340)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/util/quotapool/quotapool.go, line 258 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MAaoimm7moFKixglJmY:-MAaoimm7moFKixglJmZ:bpeid94) ([raw file](https://github.com/cockroachdb/cockroach/blob/caebc35e693afc0fe37e43b1322c7854765b395c/pkg/util/quotapool/quotapool.go#L258)):*
> ```Go
> 	var tryAgainTimer Timer
> 	var tryAgainTimerC <-chan time.Time
> 	stopTryAgainTimer := func() {
> ```

Do we need to defer this function so that we ensure that we stop the timer?
___
*[pkg/util/quotapool/quotapool.go, line 277 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MAaoJ1A7c2KLJ_qtsa4:-MAaoJ1B--XBFdMsRtRr:bdcycva) ([raw file](https://github.com/cockroachdb/cockroach/blob/caebc35e693afc0fe37e43b1322c7854765b395c/pkg/util/quotapool/quotapool.go#L277)):*
> ```Go
> 	tryAcquire := func() (fulfilled bool) {
> 		stopTryAgainTimer()
> 		fulfilled, tryAgainAfter = qp.tryAcquireOnNotify(ctx, r, n)
> ```

Could you confirm that this trio of closures all referencing each other and the timer isn't causing these variables to escape to the heap?
___
*[pkg/util/quotapool/quotapool.go, line 328 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MAbG-dc2eloHgmWtOoS:-MAbG-dd1usa-XDOnyWj:be7gmzo) ([raw file](https://github.com/cockroachdb/cockroach/blob/caebc35e693afc0fe37e43b1322c7854765b395c/pkg/util/quotapool/quotapool.go#L328)):*
> ```Go
> 	if qp.mu.q.len == 0 {
> 		var unused Resource
> 		fulfilled, unused, tryAgainAfter = r.Acquire(ctx, qp.mu.quota)
> ```

Does this take into consideration the existing waiters? For instance, if we have 10 waiters each requesting 6 quota and we replenish at a rate of 3 quota per second, will they all have timers set for 2 seconds? That seems ok and maybe even necessary to handle cancelation, but I want to confirm that the rest of this design is expecting this. 

For instance, if the second waiter calls `tryAcquire` before the first waiter, will it skip the line? That seems bad, because we could then deadlock if the quota becomes split between multiple waiters. Hopefully I'm just missing how this kind of situation is handled.
___
*[pkg/util/quotapool/time.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MAal-mh1rsIk_R6U42E:-MAal-mh1rsIk_R6U42F:b-28dn7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/a227d46ee1fbf279cc2223e0f131163fcde205a6/pkg/util/quotapool/time.go#L30)):*
> ```Go
> 	Reset(duration time.Duration)
> 	Ch() <-chan time.Time
> 	Stop() bool
> ```

minor nit: I'd move `Stop` up below `Reset` in the interface definition and down below, as they're both methods that give instructions to the timer.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50340)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/util/quotapool/quotapool.go, line 258 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MAaoimm7moFKixglJmY:-MAdnepn-daXHOVADKLx:b3z48t9) ([raw file](https://github.com/cockroachdb/cockroach/blob/caebc35e693afc0fe37e43b1322c7854765b395c/pkg/util/quotapool/quotapool.go#L258)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need to defer this function so that we ensure that we stop the timer?
</blockquote></details>

indeed, good catch, I think I killed that in a refactor at some point.
___
*[pkg/util/quotapool/quotapool.go, line 277 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MAaoJ1A7c2KLJ_qtsa4:-MAdnpC_7DcnxeohYuhS:b-mkyle8) ([raw file](https://github.com/cockroachdb/cockroach/blob/caebc35e693afc0fe37e43b1322c7854765b395c/pkg/util/quotapool/quotapool.go#L277)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could you confirm that this trio of closures all referencing each other and the timer isn't causing these variables to escape to the heap?
</blockquote></details>

```
pkg/util/quotapool/quotapool.go:352:8: func literal does not escape
pkg/util/quotapool/quotapool.go:234:9: func literal does not escape
pkg/util/quotapool/quotapool.go:263:23: func literal does not escape
pkg/util/quotapool/quotapool.go:270:24: func literal does not escape
pkg/util/quotapool/quotapool.go:280:16: func literal does not escape
```

Escape analysis on closures got reasonably good in 1.13!
___
*[pkg/util/quotapool/quotapool.go, line 328 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MAbG-dc2eloHgmWtOoS:-MAdnv487xW71GYD4YHE:b-quv6ln) ([raw file](https://github.com/cockroachdb/cockroach/blob/caebc35e693afc0fe37e43b1322c7854765b395c/pkg/util/quotapool/quotapool.go#L328)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this take into consideration the existing waiters? For instance, if we have 10 waiters each requesting 6 quota and we replenish at a rate of 3 quota per second, will they all have timers set for 2 seconds? That seems ok and maybe even necessary to handle cancelation, but I want to confirm that the rest of this design is expecting this. 

For instance, if the second waiter calls `tryAcquire` before the first waiter, will it skip the line? That seems bad, because we could then deadlock if the quota becomes split between multiple waiters. Hopefully I'm just missing how this kind of situation is handled.
</blockquote></details>

I'm not sure I follow. Only the head of the queue is ever waiting. We only attempt this fast path if there are no existing waiter (hence `qp.mu.q.len == 0`).
___
*[pkg/util/quotapool/time.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MAal-mh1rsIk_R6U42E:-MAdp2N_8XIeo0gbf-54:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a227d46ee1fbf279cc2223e0f131163fcde205a6/pkg/util/quotapool/time.go#L30)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

minor nit: I'd move `Stop` up below `Reset` in the interface definition and down below, as they're both methods that give instructions to the timer.
</blockquote></details>

done.


<!-- Sent from Reviewable.io -->

tbg commented on pkg/util/quotapool/manual_time.go:
```diff
@@ -0,0 +1,166 @@
+}
+
+func (m *ManualTime) remove(mt *manualTimer) bool {
+	// TODO(ajwerner): Consider making this a O(log N) operation instead of an
```
For a testing implementation this is hardly going to be relevant. But if you're using it for tenancy, then perhaps yes (and you're probably removing the reference to testing on this struct in some future commit).

tbg commented on pkg/util/quotapool/intpool_test.go:
```diff
@@ -813,3 +814,67 @@ func TestFreezeUnavailableCapacityPanics(t *testing.T) {
+	acq1.Release()
+	<-acquireFuncCalled
+	mt.Advance(time.Minute)
+	waitFor1(t, &called)
```
check that calledAfter is still zero?

tbg commented on pkg/util/quotapool/int_rate_impl_test.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+// NewRateLimiter defines a new RateLimiter.
+func NewRateLimiter(name string, limit Limit, burst int64, options ...Option) *RateLimiter {
```
How does the `burst` behave here? We ran into lots of trouble with stdlib `rate.Limiter` because you weren't allowed to ask more than the burst in any single acquisition (hence https://github.com/cockroachdb/cockroach/blob/76c63ded01f18aa31c08ed6a2e9cf4972b0b0459/pkg/util/limit/rate_limiter.go#L39 came into existence). The mismatch between the underlying types for `rateLimit` and `burst` was also annoying, though it's probably fine if the burst here is better behaved.

tbg replied:
PS this is also a request to clarify the semantics in the comments (chiefly `NewRateLimiter`), if you don't mind.

tbg replied:
Also, it seems that we may just remove `*LimiterBurstDisabled` when you're done here, and just use yours.

tbg commented on pkg/util/quotapool/int_rate_impl_test.go:
```diff
@@ -0,0 +1,190 @@
+var _ Resource = (*rateAlloc)(nil)
+
+func (i *rateAlloc) Merge(other Resource) {
+	panic("merge not implemented on rateAlloc")
```
Does this bother you?

tbg commented on pkg/util/quotapool/int_rate_impl_test.go:
```diff
@@ -0,0 +1,190 @@
+	rl    *RateLimiter
+}
+
+// Return returns the RateAlloc to the RateLimiter. It is not safe to call any
```
There's no technical problem with later adding code that returns quota to the pool that wasn't initially in it, right? Thinking about Ben's proposal. In that case, we probably also want to keep the quota even if it overflows the bucket (i.e. exceeds burst).

tbg replied:
(but one step at a time).

tbg commented on pkg/util/quotapool/int_rate_impl_test.go:
```diff
@@ -0,0 +1,190 @@
+	}
+	if i.want > int64(r.cur) {
+		// Compute the time it will take for r.cur to get to the needed capacity.
+		delta := float64(i.want) - r.cur
```
Can there be cases in which `i.want > int64(r.cur)` but `float64(i.want) < r.cur`?  The conversions here just make me anxious. Can this become

```go
if delta := float64(i.want) - r.cur; delta > 0 {
...
}
```

tbg commented on pkg/util/quotapool/int_rate_impl_test.go:
```diff
@@ -0,0 +1,190 @@
+		timeDelta := time.Duration(seconds * float64(time.Second))
+		return false, nil, timeDelta
+	}
+	r.cur -= float64(i.want)
```
This line also looks saner with the rewrite above

tbg commented on pkg/util/quotapool/int_rate_test.go:
```diff
@@ -0,0 +1,114 @@
+		quotapool.WithTimeSource(mt))
+	ctx := context.Background()
+
+	// Consume all of the burst over a number of requests.
```
Asking for more than the burst should also get a test.

tbg reviewed (approved):
My main question is whether the semantics of the burst are useful. They were a PITA last time I touched them with @lunevalex. What's the rationale for flat-out refusing acquisitions that need more than the burst in a rate limiter? If there's a burst of 1 and I'm asking for 2, I will have to wait, seems fine, better than forcing every user of this to make sure they *never* pass anything larger than the burst, which is just going to lead to surprises and then awkward workarounds like `*LimiterBurstDisabled` (which has no fairness).

nvb reviewed (commented):
Planning on giving the `RateLimiter` itself another pass once Tobi's comments are resolved. He seems to have stronger opinions on that than I did.

> Reviewed 1 of 8 files at r5, 2 of 2 files at r6, 1 of 1 files at r7, 1 of 4 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-:-MAhY3d8Btj03kGTngg_:bsplptp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @nvanbenschoten)

___
*[pkg/util/quotapool/quotapool.go, line 277 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MAaoJ1A7c2KLJ_qtsa4:-MAhXN2l8--DTO1eX4qy:bdnffi0) ([raw file](https://github.com/cockroachdb/cockroach/blob/caebc35e693afc0fe37e43b1322c7854765b395c/pkg/util/quotapool/quotapool.go#L277)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

```
pkg/util/quotapool/quotapool.go:352:8: func literal does not escape
pkg/util/quotapool/quotapool.go:234:9: func literal does not escape
pkg/util/quotapool/quotapool.go:263:23: func literal does not escape
pkg/util/quotapool/quotapool.go:270:24: func literal does not escape
pkg/util/quotapool/quotapool.go:280:16: func literal does not escape
```

Escape analysis on closures got reasonably good in 1.13!
</blockquote></details>

Nice! We've come so far.
___
*[pkg/util/quotapool/quotapool.go, line 328 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MAbG-dc2eloHgmWtOoS:-MAhXSUjAmtcI-OC6Ak0:b-mflja8) ([raw file](https://github.com/cockroachdb/cockroach/blob/caebc35e693afc0fe37e43b1322c7854765b395c/pkg/util/quotapool/quotapool.go#L328)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'm not sure I follow. Only the head of the queue is ever waiting. We only attempt this fast path if there are no existing waiter (hence `qp.mu.q.len == 0`).
</blockquote></details>

Ah, I totally missed that only the head of the queue will have an active timer. I think the `return ..., tryAgainAfter, ...` below confused me, but I see why that's there now. That property is important enough and such a nice simplification to the logic in `Acquire` that I'd write it out in a comment somewhere.
___
*[pkg/util/quotapool/quotapool.go, line 256 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MAhUtOQEx4k9nubk6TO:-MAhUtOR7Pz6KyuH-nqV:b61xqm6) ([raw file](https://github.com/cockroachdb/cockroach/blob/62bf4c398bc6af8df84d1ad5e909356d06d9281a/pkg/util/quotapool/quotapool.go#L256)):*
> ```Go
> 	// Set up the infrastructure to deal with rate-limiter style pools which
> 	// retry after the passage of time.
> 	var (
> ```

Not that it's a bike worth shedding or a shed worth biking, but the more I see it, the more I'm :thumbsdown: on combing `var` blocks with the closure-heavy style we have here. For one, it causes two extra outdents for all of the business logic in the function, which makes the logic easier to miss while skimming. It also gets confusing when the vars start to depend on one another because `var` blocks don't immediately indicate variable definition in program order. Dependent vars need to be declared in order in functions themselves, but that's not true when in the global scope and that's an easy rule to forget as a reader, so I end up looking all over for variable defs instead of just looking up.

Do you have a strong affection for this style?
___
*[pkg/util/quotapool/time.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MAal-mh1rsIk_R6U42E:-MAhXxTo4CXOOGFGR9bb:b-8r63sw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a227d46ee1fbf279cc2223e0f131163fcde205a6/pkg/util/quotapool/time.go#L30)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

done.
</blockquote></details>

And the implementation :smiley: 


<!-- Sent from Reviewable.io -->

Commit:
util/quotapool: adjust Resource interface

The Resource object in practice is always modified in place and then returned
from Acquire. There was no real reason to retain the symmetry between the
Resource held by the QuotaPool and value that we merge into it. For the
IntPool, the symmetry happened for free because we'd hand out the same thing
we'd track but that becomes untrue when we implement rate limiters.

Release note: None

ajwerner reviewed (commented):
> My main question is whether the semantics of the burst are useful. They were a PITA last time I touched them with @lunevalex. What's the rationale for flat-out refusing acquisitions that need more than the burst in a rate limiter? If there's a burst of 1 and I'm asking for 2, I will have to wait, seems fine, better than forcing every user of this to make sure they never pass anything larger than the burst, which is just going to lead to surprises and then awkward workarounds like *LimiterBurstDisabled (which has no fairness).

Sure, I'm on board with that behavior being crappy. I mostly did it because it's what https://godoc.org/golang.org/x/time/rate#Limiter.WaitN does and this implementation was just to test that `tryAgainAfter` works as I expected it to, not necessarily to make a wonderful library.  I'm on board with changing that perspective. I changed the semantics to be like what I did for the `tenantrate.Limiter`. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-:-MAiF8mb8vaBGY_YTsDj:bapse1p)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[pkg/util/quotapool/int_rate_impl_test.go, line 36 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MA1gyWH8hkPL0fXzRZR-r8-36:-MAiFpi8CuJvWNZz89cu:bo1b2xx) ([raw file](https://github.com/cockroachdb/cockroach/blob/18805ddc8f4a74c27974b2062f8055da04975d39/pkg/util/quotapool/int_rate_impl_test.go#L36)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Also, it seems that we may just remove `*LimiterBurstDisabled` when you're done here, and just use yours.
</blockquote></details>

Fair. I updated the comment and changed the burst behavior to reflect that in use in the `tenantrate.Limiter` - if you want to acquire more than the burst, you can, but only if the bucket is totally full and it puts the bucket into debt. I still need to implement adjusting the rate but it should be trivial. We can leave that adoption for another PR.
___
*[pkg/util/quotapool/int_rate_impl_test.go, line 99 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MA1gyWH8hkPL0fXzRZR-r8-99:-MAiGGcsBOCE0maE0Uwt:btzflzw) ([raw file](https://github.com/cockroachdb/cockroach/blob/18805ddc8f4a74c27974b2062f8055da04975d39/pkg/util/quotapool/int_rate_impl_test.go#L99)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

(but one step at a time).
</blockquote></details>

Correct, it's fine. 
___
*[pkg/util/quotapool/int_rate_impl_test.go, line 117 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MA1gyWH8hkPL0fXzRZR-r8-117:-MAiGNpG4mQos_JGw2aV:b-sywn2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/18805ddc8f4a74c27974b2062f8055da04975d39/pkg/util/quotapool/int_rate_impl_test.go#L117)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Does this bother you?
</blockquote></details>

Yes this and reflecting on the other users of this interface indicates that I've got the abstractions wrong. I've added a commit to the front of this PR to deal with the shape of the `Resource` interface. I think it's better. As long as resource gets the type it expects, everything is happy.
___
*[pkg/util/quotapool/int_rate_impl_test.go, line 164 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MA1gyWH8hkPL0fXzRZR-r8-164:-MAiQ3B25UuNgOYKyiNU:b7btn1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/18805ddc8f4a74c27974b2062f8055da04975d39/pkg/util/quotapool/int_rate_impl_test.go#L164)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Can there be cases in which `i.want > int64(r.cur)` but `float64(i.want) < r.cur`?  The conversions here just make me anxious. Can this become

```go
if delta := float64(i.want) - r.cur; delta > 0 {
...
}
```
</blockquote></details>

Done. Though almost certain the answer is no.  Now, however, I am slightly worried about a case where delta > 0 and the timeDelta becomes zero. I think I made that much harder by moving the multiplication by `time.Second` up but with the right set of values, it is possible. I'm thinking I should, as a safeguard add a 1 here if timeDelta is zero just to make it all happy. Returning a zero would be, well, very bad. The other option is to add a single nanosecond always but that will make testing more annoying. 

I added a test for this. 
___
*[pkg/util/quotapool/int_rate_impl_test.go, line 169 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MA1gyWH8hkPL0fXzRZR-r8-169:-MAiryUl82qVqbE2m9lp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/18805ddc8f4a74c27974b2062f8055da04975d39/pkg/util/quotapool/int_rate_impl_test.go#L169)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This line also looks saner with the rewrite above
</blockquote></details>

Done.
___
*[pkg/util/quotapool/int_rate_test.go, line 34 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MA1gyWH8hkPL0fXzRZS-r8-34:-MAiyiE52FjYl1EoG2cD:b2cq2l) ([raw file](https://github.com/cockroachdb/cockroach/blob/18805ddc8f4a74c27974b2062f8055da04975d39/pkg/util/quotapool/int_rate_test.go#L34)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Asking for more than the burst should also get a test.
</blockquote></details>

Tested the new policy. 
___
*[pkg/util/quotapool/intpool_test.go, line 876 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MA1gyWGVDq0AGUg2dVv-r7-876:-MAiz4rf3CvTB9P80zXV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7ed2263c2b297eb05c172d098baed1127eebf5c/pkg/util/quotapool/intpool_test.go#L876)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

check that calledAfter is still zero?
</blockquote></details>

Done.
___
*[pkg/util/quotapool/manual_time.go, line 87 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MA1gyWGVDq0AGUg2dVt-r6-87:-MAiH72XB_1RtDbLCW0B:b-kxzrwq) ([raw file](https://github.com/cockroachdb/cockroach/blob/599c60f26c2859833b83803327403cdb4b1b09b7/pkg/util/quotapool/manual_time.go#L87)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

For a testing implementation this is hardly going to be relevant. But if you're using it for tenancy, then perhaps yes (and you're probably removing the reference to testing on this struct in some future commit).
</blockquote></details>

This just exists for testing there too. 
___
*[pkg/util/quotapool/quotapool.go, line 328 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MAbG-dc2eloHgmWtOoS:-MAj-HQbAaDI4hsXTL_l:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/caebc35e693afc0fe37e43b1322c7854765b395c/pkg/util/quotapool/quotapool.go#L328)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Ah, I totally missed that only the head of the queue will have an active timer. I think the `return ..., tryAgainAfter, ...` below confused me, but I see why that's there now. That property is important enough and such a nice simplification to the logic in `Acquire` that I'd write it out in a comment somewhere.
</blockquote></details>

Done.
___
*[pkg/util/quotapool/quotapool.go, line 256 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MAhUtOQEx4k9nubk6TO:-MAiHEpP8ctVk3THbwOm:b9mqald) ([raw file](https://github.com/cockroachdb/cockroach/blob/62bf4c398bc6af8df84d1ad5e909356d06d9281a/pkg/util/quotapool/quotapool.go#L256)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Not that it's a bike worth shedding or a shed worth biking, but the more I see it, the more I'm :thumbsdown: on combing `var` blocks with the closure-heavy style we have here. For one, it causes two extra outdents for all of the business logic in the function, which makes the logic easier to miss while skimming. It also gets confusing when the vars start to depend on one another because `var` blocks don't immediately indicate variable definition in program order. Dependent vars need to be declared in order in functions themselves, but that's not true when in the global scope and that's an easy rule to forget as a reader, so I end up looking all over for variable defs instead of just looking up.

Do you have a strong affection for this style?
</blockquote></details>

:shrug: no strong feelings here. 
___
*[pkg/util/quotapool/time.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MAal-mh1rsIk_R6U42E:-MAj-JssAdZMuGfgqjbE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a227d46ee1fbf279cc2223e0f131163fcde205a6/pkg/util/quotapool/time.go#L30)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

And the implementation :smiley: 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50340)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[pkg/util/quotapool/manual_time.go, line 87 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50340#-MA1gyWGVDq0AGUg2dVt-r6-87:-MAl4QH-BvJEquAic5aL:b-fhhsnb) ([raw file](https://github.com/cockroachdb/cockroach/blob/599c60f26c2859833b83803327403cdb4b1b09b7/pkg/util/quotapool/manual_time.go#L87)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This just exists for testing there too. 
</blockquote></details>

Drive-by fixed it for no good reason. 


<!-- Sent from Reviewable.io -->

ajwerner commented:
I think this is in a good enough place. I'm going to merge it to simplify the `tenantrate` PR. TFTR!

bors r+

ajwerner commented:
bors r+

Timeout in acceptance test
